Add route to count comments, add count to GET request result

smbkr opened the pull request:
Addresses #78.

Get a count for a url with `host:port/count?url=someurl.with/comments`.

Also adds count to comment GET request. Successful response for /get is now:
`{"success":true,"message":"","comments":[...],"count":5}`

Response for `/count`:
`{"success":true,"message":"","count":1}`

GaiusMartius commented on http.go:
```diff
@@ -17,11 +17,12 @@ type resultContainer struct {
 	Success  bool      `json:"success"`
 	Message  string    `json:"message"`
 	Comments []Comment `json:"comments,omitempty"`
+	Count    *int       `json:"count,omitempty"`
```
Int is simple data type so it should be a value.

You can leave out the 'omitempty' as well. Count will default to 0, so if there are no comments then the default value is reasonable.

smbkr replied:
Actually I did it like you suggest originally, but then changed to this since I wanted the count to be left out for a POST request, and this was a way to do so, since the count could also be nil, but having given it some thought, it makes sense to just return the updated count after a successful POST. So I will update the PR to incorporate that & your other suggestions. Thanks for the review!

adtac replied:
+1 for `omitempty`

adtac replied:
but why is this a pointer to an int?

GaiusMartius commented on http.go:
```diff
@@ -17,11 +17,12 @@ type resultContainer struct {
 func (res *resultContainer) render(w http.ResponseWriter) {
 	if res == nil {
-		res = &resultContainer{false, "Some internal error occurred", nil}
+		res = &resultContainer{false, "Some internal error occurred", nil, nil}
```
With Count as a value instead of a pointer the initializer can be left unchanged. Count will assume the reasonable default of 0

adtac replied:
Yep, we can have this as a value. Or better yet, initialize this as 0.

If you're worried about `omitempty` removing it because it's 0, we can handle that in front-end.

GaiusMartius commented on http.go:
```diff
@@ -82,5 +83,34 @@ func GetCommentsHandler(w http.ResponseWriter, r *http.Request) {
 		Emit(err)
 	}
 	result.Comments = comments
+	count := int(len(comments))
```
len() will already produce an int so, no need for a cast here

GaiusMartius commented on http.go:
```diff
@@ -82,5 +83,34 @@ func GetCommentsHandler(w http.ResponseWriter, r *http.Request) {
+	}
+
+	result.Success = true
+	count = int(count)
```
You can remove this line.

adtac commented on comment.go:
```diff
@@ -74,3 +75,20 @@ func getComments(url string) ([]Comment, error) {
+	`
+	err := db.QueryRow(statement, url).Scan(&count)
+	switch {
+	case err == sql.ErrNoRows:
```
I didn't know Go allowed cases like this in `switch` :3 

I think a `if(condition) { ... } else if(condition) { ... } else { ... }` would be simpler here

adtac commented on comment.go:
```diff
@@ -74,3 +75,20 @@ func getComments(url string) ([]Comment, error) {
+	}
+
+	return count, nil
+}
```
Newline at the end of the file please

adtac commented on comment.go:
```diff
@@ -74,3 +75,20 @@ func getComments(url string) ([]Comment, error) {
+	case err == sql.ErrNoRows:
+		count = 0
+	case err != nil:
+		log.Println(err)
```
you'll want to `return 0, err` here

adtac commented:
unack https://github.com/adtac/commento/pull/85/commits/befd5b3928e296e86744825e07854df0a844c09c

adtac commented:
@smbkr ping here too :)

smbkr commented:
@adtac so sorry! been massively busy lately! I'd still like to fix this & my other PR with the suggested changes, will take a look at my lunch today & will also attempt merging master into this branch so that conflicts can be avoided.

adtac commented:
@smbkr no problem, it's open source; everyone is just a volunteer.

And thanks for getting this going again!

smbkr commented:
@adtac I've just pushed some commits which I believe address all the issues raised - I've removed unnecessary casts and pointers, changed the switch in the countComments method to a standard if/else, and also added the updated count when creating a comment.

I've also removed `omitempty` from the count, as the value `0` is treated as empty, and we want this to be returned in the response when the actual count is zero. Previously I had used this as I was not returning a count when creating a comment, so I wanted that key to be omitted, however I think it makes sense to return the updated count after a new comment is created. This was a count is returned with `/get` and `/create`, and can be retrieved on its own with `/count`. What do you think?

adtac commented:
@smbkr Neat stuff, thanks!

Just one tiny thing: please use rebases instead of merging from master. This helps us maintain a linear history and they [are good for a number of reasons](https://stackoverflow.com/questions/20348629/what-are-advantages-of-keeping-linear-history-in-git).

Alternatively, you can hand over the PR to me and I'll cherry-pick your commits (author information will be intact) in a larger PR. I'll probably be doing this anyway since there are multiple huge PRs close to merging, so it's really not a problem for me.

smbkr commented:
Cheers @adtac - I'm happy for you to cherry pick. Same goes for the other PR. I'm used to just merging master into my working branch since that's the way we do it at my work, however I can see the benefit of rebasing instead. I'll try that next time!

adtac commented:
Thanks for letting me know, @smbkr :)

I'll comment here when they get merged through a cherry-pick (just as a heads-up).
